stablehlo/reference/Ops.cpp

```diff
@@ -94,6 +94,34 @@ Tensor evalClampOp(const Tensor &min, const Tensor &operand, const Tensor &max,
   return result;
 }
 
+Tensor evalCompareOp(const Tensor &lhs, const Tensor &rhs,
```

Comment:
IMO the implementation can be simplified a lot if we define operator overloads for `<=` , `<`, `>=` and `>` for reuse. 


Here is the suggestion
```
Tensor evalCompareOp(const Tensor &lhs, const Tensor &rhs,
                     ComparisonDirection comparisonDirection,
                     std::optional<ComparisonType> compareType,
                     TensorType resultType) {
  Tensor result(resultType);
  auto elementTy = result.getType().dyn_cast<TensorType>().getElementType();
  auto comparisonType = ComparisonType::NOTYPE;
  if (compareType && *compareType != ComparisonType::NOTYPE) {
    comparisonType = *compareType;
  } else if (isSupportedUnsignedIntegerType(elementTy) ||
             isSupportedBooleanType(elementTy)) {
    comparisonType = ComparisonType::UNSIGNED;
  } else if (isSupportedSignedIntegerType(elementTy)) {
    comparisonType = ComparisonType::SIGNED;
  } else if (isSupportedFloatType(elementTy) ||
             isSupportedComplexType(elementTy)) {
    comparisonType = ComparisonType::FLOAT;
  } else {
    report_fatal_error(invalidArgument("Unsupported element type: %s",
                                       debugString(elementTy).c_str()));
  }

  for (auto it = result.index_begin(); it != result.index_end(); ++it) {
    bool cmpResult = false;
    switch (comparisonDirection) {
      case ComparisonDirection::EQ: {
        if (comparisonType == ComparisonType::TOTALORDER)
          cmpResult = lhs.get(*it).getFloatValue().bitcastToAPInt() ==
                      rhs.get(*it).getFloatValue().bitcastToAPInt();
        else
          cmpResult = lhs.get(*it) == rhs.get(*it);
        break;
      }
      case ComparisonDirection::NE: {
        if (comparisonType == ComparisonType::TOTALORDER)
          cmpResult = lhs.get(*it).getFloatValue().bitcastToAPInt() !=
                      rhs.get(*it).getFloatValue().bitcastToAPInt();
        else
          cmpResult = lhs.get(*it) != rhs.get(*it);
        break;
      }
      case ComparisonDirection::GE: {
        if (comparisonType == ComparisonType::TOTALORDER) {
          auto lhsFloat = lhs.get(*it).getFloatValue();
          auto rhsFloat = rhs.get(*it).getFloatValue();
          if (lhsFloat.isNegative() ^ rhsFloat.isNegative())
            cmpResult = rhsFloat.isNegative();
          else
            cmpResult =
                lhsFloat.bitcastToAPInt().uge(rhsFloat.bitcastToAPInt());
        } else {
          cmpResult = lhs.get(*it) >= rhs.get(*it);
        }
        break;
      }
      case ComparisonDirection::GT: {
        if (comparisonType == ComparisonType::TOTALORDER) {
          auto lhsFloat = lhs.get(*it).getFloatValue();
          auto rhsFloat = rhs.get(*it).getFloatValue();
          if (lhsFloat.isNegative() ^ rhsFloat.isNegative())
            cmpResult = rhsFloat.isNegative();
          else
            cmpResult =
                lhsFloat.bitcastToAPInt().ugt(rhsFloat.bitcastToAPInt());
        } else {
          cmpResult = lhs.get(*it) > rhs.get(*it);
        }
        break;
      }
      case ComparisonDirection::LE: {
        if (comparisonType == ComparisonType::TOTALORDER) {
          auto lhsFloat = lhs.get(*it).getFloatValue();
          auto rhsFloat = rhs.get(*it).getFloatValue();
          if (lhsFloat.isNegative() ^ rhsFloat.isNegative())
            cmpResult = lhsFloat.isNegative();
          else
            cmpResult =
                lhsFloat.bitcastToAPInt().ule(rhsFloat.bitcastToAPInt());
        } else {
          cmpResult = lhs.get(*it) <= rhs.get(*it);
        }
        break;
      }
      case ComparisonDirection::LT: {
        if (comparisonType == ComparisonType::TOTALORDER) {
          auto lhsFloat = lhs.get(*it).getFloatValue();
          auto rhsFloat = rhs.get(*it).getFloatValue();
          if (lhsFloat.isNegative() ^ rhsFloat.isNegative())
            cmpResult = lhsFloat.isNegative();
          else
            cmpResult =
                lhsFloat.bitcastToAPInt().ult(rhsFloat.bitcastToAPInt());
        } else {
          cmpResult = lhs.get(*it) < rhs.get(*it);
        }
        break;
      }
    }
    result.set(*it, Element(elementTy, cmpResult));
  }
  return result;
}
```

and in `Element.cpp` we can have the overloads for `Element::operator<`. 

```
bool Element::operator<(const Element &other) const {
  Type type = other.getType();
  if (type_ != type)
    report_fatal_error(invalidArgument("Element types don't match: %s vs %s",
                                       debugString(type_).c_str(),
                                       debugString(type).c_str()));

  if (isSupportedIntegerType(type)) {
    auto intLhs = getIntegerValue();
    auto intRhs = other.getIntegerValue();
    return isSupportedSignedIntegerType(type) ? intLhs.slt(intRhs)
                                              : intLhs.ult(intRhs);
  }

  if (isSupportedBooleanType(type)) {
    auto boolLhs = getBooleanValue();
    auto boolRhs = other.getBooleanValue();
    return boolLhs < boolRhs;
  }

  if (isSupportedFloatType(type)) {
    auto floatLhs = getFloatValue();
    auto floatRhs = other.getFloatValue();
    return floatLhs.compare(floatRhs) == APFloat::cmpLessThan;
  }

  if (isSupportedComplexType(type)) {
    // TODO
  }

  report_fatal_error(invalidArgument("Unsupported element type: %s",
                                     debugString(type).c_str()));
}

bool Element::operator<=(const Element &other) const {
  return !(*this < other) || (*this == other);
}

bool Element::operator>(const Element &other) const {
  return !(*this <= other);
}

bool Element::operator>=(const Element &other) const {
  return !(*this < other);
}
```

Reply:
Thanks for the suggestion, I've borrowed majority of this, except that we still need to define our own function for operator> to avoid accidentally negating NaNs (e.g. !(NaN <= NaN) is true when it should be false).

Reply:
>> need to define our own function for operator> to avoid accidentally negating NaNs
That is right. Thanks!